File: microgpt.dart
/// The most atomic way to train and inference a GPT in pure, dependency-free Dart.
/// This file is the complete algorithm.
/// Everything else is just efficiency.
///
/// @karpathy (original Python), converted to Dart

import 'dart:io';
import 'dart:math';

// --- RNG (xoshiro256** seeded with 42) ---
class Rng {
  late int s0, s1, s2, s3; // 64-bit state

  Rng(int seed) {
    // SplitMix64 to seed xoshiro
    int sm = seed;
    int smNext() {
      sm = (sm + 0x9e3779b97f4a7c15) & 0xFFFFFFFFFFFFFFFF;
      int z = sm;
      z = ((z ^ (z >> 30)) * 0xbf58476d1ce4e5b9) & 0xFFFFFFFFFFFFFFFF;
      z = ((z ^ (z >> 27)) * 0x94d049bb133111eb) & 0xFFFFFFFFFFFFFFFF;
      return (z ^ (z >> 31)) & 0xFFFFFFFFFFFFFFFF;
    }

    s0 = smNext();
    s1 = smNext();
    s2 = smNext();
    s3 = smNext();
  }

  int _rotl(int x, int k) =>
      ((x << k) | ((x & 0xFFFFFFFFFFFFFFFF) >>> (64 - k))) & 0xFFFFFFFFFFFFFFFF;

  int nextU64() {
    final result =
        (_rotl((s1 * 5) & 0xFFFFFFFFFFFFFFFF, 7) * 9) & 0xFFFFFFFFFFFFFFFF;
    final t = (s1 << 17) & 0xFFFFFFFFFFFFFFFF;
    s2 ^= s0;
    s3 ^= s1;
    s1 ^= s2;
    s0 ^= s3;
    s2 ^= t;
    s3 = _rotl(s3, 45);
    return result;
  }

  double nextF64() => (nextU64() >>> 11) / (1 << 53);

  double gauss(double mean, double std) {
    final u1 = max(nextF64(), 1e-30);
    final u2 = nextF64();
    return mean + std * sqrt(-2.0 * log(u1)) * cos(2.0 * pi * u2);
  }

  void shuffle<T>(List<T> list) {
    for (int i = list.length - 1; i > 0; i--) {
      final j = nextU64() % (i + 1);
      final tmp = list[i];
      list[i] = list[j];
      list[j] = tmp;
    }
  }

  int weightedChoice(List<double> weights) {
    final total = weights.reduce((a, b) => a + b);
    double r = nextF64() * total;
    for (int i = 0; i < weights.length; i++) {
      r -= weights[i];
      if (r <= 0) return i;
    }
    return weights.length - 1;
  }
}

// --- Let there be Autograd ---
class Value {
  double data; // scalar value of this node calculated during forward pass
  double grad = 0.0; // derivative of the loss w.r.t. this node
  final List<Value> children; // children of this node in the computation graph
  final List<double>
  localGrads; // local derivative of this node w.r.t. its children

  Value(this.data, {List<Value>? children, List<double>? localGrads})
    : children = children ?? [],
      localGrads = localGrads ?? [];

  Value operator +(Value other) =>
      Value(data + other.data, children: [this, other], localGrads: [1.0, 1.0]);

  Value addScalar(double s) => this + Value(s);

  Value operator *(Value other) => Value(
    data * other.data,
    children: [this, other],
    localGrads: [other.data, data],
  );

  Value mulScalar(double s) => this * Value(s);

  Value vpow(double exp) => Value(
    pow(data, exp).toDouble(),
    children: [this],
    localGrads: [exp * pow(data, exp - 1).toDouble()],
  );

  Value vlog() => Value(log(data), children: [this], localGrads: [1.0 / data]);

  Value vexp() {
    final e = exp(data);
    return Value(e, children: [this], localGrads: [e]);
  }

  Value relu() =>
      Value(max(0, data), children: [this], localGrads: [data > 0 ? 1.0 : 0.0]);

  Value operator -() => mulScalar(-1.0);

  Value operator -(Value other) => this + (-other);

  Value operator /(Value other) => this * other.vpow(-1.0);

  Value divScalar(double s) => mulScalar(1.0 / s);

  void backward() {
    final topo = <Value>[];
    final visited = <Value>{};

    void buildTopo(Value v) {
      if (visited.contains(v)) return;
      visited.add(v);
      for (final child in v.children) {
        buildTopo(child);
      }
      topo.add(v);
    }

    buildTopo(this);
    grad = 1.0;
    for (final v in topo.reversed) {
      for (int i = 0; i < v.children.length; i++) {
        v.children[i].grad += v.localGrads[i] * v.grad;
      }
    }
  }
}

// --- Utility ---
Value sumVals(List<Value> v) {
  var s = Value(0.0);
  for (final x in v) s = s + x;
  return s;
}

// --- Model functions ---
List<Value> linear(List<Value> x, List<List<Value>> w) {
  return w.map((wo) {
    var s = Value(0.0);
    for (int i = 0; i < wo.length; i++) {
      s = s + wo[i] * x[i];
    }
    return s;
  }).toList();
}

List<Value> softmax(List<Value> logits) {
  final maxVal = logits.map((v) => v.data).reduce(max);
  final exps = logits.map((v) => (v - Value(maxVal)).vexp()).toList();
  final total = sumVals(exps);
  return exps.map((e) => e / total).toList();
}

List<Value> rmsnorm(List<Value> x) {
  var ms = Value(0.0);
  for (final xi in x) ms = ms + xi * xi;
  ms = ms.divScalar(x.length.toDouble());
  final scale = (ms.addScalar(1e-5)).vpow(-0.5);
  return x.map((xi) => xi * scale).toList();
}

// --- Main ---
void main() {
  final rng = Rng(42);

  // Let there be an input dataset
  final file = File('input.txt');
  if (!file.existsSync()) {
    print('Error: input.txt not found. Please download it first.');
    exit(1);
  }
  final content = file.readAsStringSync();
  final docs = content
      .split('\n')
      .map((l) => l.trim())
      .where((l) => l.isNotEmpty)
      .toList();
  rng.shuffle(docs);
  print('num docs: ${docs.length}');

  // Let there be a Tokenizer
  final charSet = <String>{};
  for (final d in docs) {
    for (int i = 0; i < d.length; i++) {
      charSet.add(d[i]);
    }
  }
  final uchars = charSet.toList()..sort();
  final bos = uchars.length;
  final vocabSize = uchars.length + 1;
  print('vocab size: $vocabSize');

  int charIndex(String c) => uchars.indexOf(c);

  // Initialize the parameters
  const nEmbd = 16;
  const nHead = 4;
  const nLayer = 1;
  const blockSize = 16;
  const headDim = nEmbd ~/ nHead;

  List<List<Value>> makeMatrix(int nout, int nin) {
    return List.generate(
      nout,
      (_) => List.generate(nin, (_) => Value(rng.gauss(0, 0.08))),
    );
  }

  final wte = makeMatrix(vocabSize, nEmbd);
  final wpe = makeMatrix(blockSize, nEmbd);
  final lmHead = makeMatrix(vocabSize, nEmbd);

  // Per-layer weights
  final layerAttnWq = <List<List<Value>>>[];
  final layerAttnWk = <List<List<Value>>>[];
  final layerAttnWv = <List<List<Value>>>[];
  final layerAttnWo = <List<List<Value>>>[];
  final layerMlpFc1 = <List<List<Value>>>[];
  final layerMlpFc2 = <List<List<Value>>>[];

  for (int i = 0; i < nLayer; i++) {
    layerAttnWq.add(makeMatrix(nEmbd, nEmbd));
    layerAttnWk.add(makeMatrix(nEmbd, nEmbd));
    layerAttnWv.add(makeMatrix(nEmbd, nEmbd));
    layerAttnWo.add(makeMatrix(nEmbd, nEmbd));
    layerMlpFc1.add(makeMatrix(4 * nEmbd, nEmbd));
    layerMlpFc2.add(makeMatrix(nEmbd, 4 * nEmbd));
  }

  // Flatten params
  final params = <Value>[];
  for (final row in wte) for (final p in row) params.add(p);
  for (final row in wpe) for (final p in row) params.add(p);
  for (final row in lmHead) for (final p in row) params.add(p);
  for (int i = 0; i < nLayer; i++) {
    for (final mat in [
      layerAttnWq[i],
      layerAttnWk[i],
      layerAttnWv[i],
      layerAttnWo[i],
      layerMlpFc1[i],
      layerMlpFc2[i],
    ]) {
      for (final row in mat) for (final p in row) params.add(p);
    }
  }
  print('num params: ${params.length}');

  // GPT function
  List<Value> gpt(
    int tokenId,
    int posId,
    List<List<List<Value>>> keys,
    List<List<List<Value>>> values,
  ) {
    final tokEmb = wte[tokenId];
    final posEmb = wpe[posId];
    var x = List.generate(nEmbd, (i) => tokEmb[i] + posEmb[i]);
    x = rmsnorm(x);

    for (int li = 0; li < nLayer; li++) {
      // 1) Multi-head attention block
      final xResidual = List<Value>.from(x);
      x = rmsnorm(x);
      final q = linear(x, layerAttnWq[li]);
      final k = linear(x, layerAttnWk[li]);
      final v = linear(x, layerAttnWv[li]);
      keys[li].add(k);
      values[li].add(v);
      final xAttn = <Value>[];
      for (int h = 0; h < nHead; h++) {
        final hs = h * headDim;
        final qH = q.sublist(hs, hs + headDim);
        final kH = keys[li].map((ki) => ki.sublist(hs, hs + headDim)).toList();
        final vH = values[li]
            .map((vi) => vi.sublist(hs, hs + headDim))
            .toList();
        final scale = sqrt(headDim.toDouble());
        final attnLogits = List.generate(kH.length, (t) {
          var dot = Value(0.0);
          for (int j = 0; j < headDim; j++) dot = dot + qH[j] * kH[t][j];
          return dot.divScalar(scale);
        });
        final attnWeights = softmax(attnLogits);
        for (int j = 0; j < headDim; j++) {
          var s = Value(0.0);
          for (int t = 0; t < vH.length; t++) s = s + attnWeights[t] * vH[t][j];
          xAttn.add(s);
        }
      }
      x = linear(xAttn, layerAttnWo[li]);
      for (int i = 0; i < nEmbd; i++) x[i] = x[i] + xResidual[i];
      // 2) MLP block
      final xResidual2 = List<Value>.from(x);
      x = rmsnorm(x);
      x = linear(x, layerMlpFc1[li]);
      x = x.map((xi) => xi.relu()).toList();
      x = linear(x, layerMlpFc2[li]);
      for (int i = 0; i < nEmbd; i++) x[i] = x[i] + xResidual2[i];
    }

    return linear(x, lmHead);
  }

  // Let there be Adam
  const learningRate = 0.01;
  const beta1 = 0.85;
  const beta2 = 0.99;
  const epsAdam = 1e-8;
  final mBuf = List.filled(params.length, 0.0);
  final vBuf = List.filled(params.length, 0.0);

  // Repeat in sequence
  const numSteps = 1000;
  for (int step = 0; step < numSteps; step++) {
    // Take single document, tokenize it
    final doc = docs[step % docs.length];
    final tokens = <int>[bos];
    for (int i = 0; i < doc.length; i++) tokens.add(charIndex(doc[i]));
    tokens.add(bos);
    final n = min(blockSize, tokens.length - 1);

    // Forward
    final keys = List.generate(nLayer, (_) => <List<Value>>[]);
    final vals = List.generate(nLayer, (_) => <List<Value>>[]);
    final losses = <Value>[];
    for (int posId = 0; posId < n; posId++) {
      final tokenId = tokens[posId];
      final targetId = tokens[posId + 1];
      final logits = gpt(tokenId, posId, keys, vals);
      final probs = softmax(logits);
      final lossT = -(probs[targetId].vlog());
      losses.add(lossT);
    }
    final loss = sumVals(losses).divScalar(n.toDouble());

    // Backward
    loss.backward();

    // Adam optimizer update
    final lrT = learningRate * (1.0 - step / numSteps);
    for (int i = 0; i < params.length; i++) {
      final p = params[i];
      mBuf[i] = beta1 * mBuf[i] + (1 - beta1) * p.grad;
      vBuf[i] = beta2 * vBuf[i] + (1 - beta2) * p.grad * p.grad;
      final mHat = mBuf[i] / (1.0 - pow(beta1, step + 1));
      final vHat = vBuf[i] / (1.0 - pow(beta2, step + 1));
      p.data -= lrT * mHat / (sqrt(vHat) + epsAdam);
      p.grad = 0;
    }

    final s1 = (step + 1).toString().padLeft(4);
    final s2 = numSteps.toString().padLeft(4);
    print('step $s1 / $s2 | loss ${loss.data.toStringAsFixed(4)}');
  }

  // Inference
  const temperature = 0.5;
  print('\n--- inference (new, hallucinated names) ---');
  for (int sampleIdx = 0; sampleIdx < 20; sampleIdx++) {
    final keys = List.generate(nLayer, (_) => <List<Value>>[]);
    final vals = List.generate(nLayer, (_) => <List<Value>>[]);
    var tokenId = bos;
    final sample = <String>[];
    for (int posId = 0; posId < blockSize; posId++) {
      final logits = gpt(tokenId, posId, keys, vals);
      final tempLogits = logits.map((l) => l.divScalar(temperature)).toList();
      final probs = softmax(tempLogits);
      final weights = probs.map((p) => p.data).toList();
      tokenId = rng.weightedChoice(weights);
      if (tokenId == bos) break;
      sample.add(uchars[tokenId]);
    }
    final idx = (sampleIdx + 1).toString().padLeft(2);
    print('sample $idx: ${sample.join()}');
  }
}

File: microgpt.js
/**
 * The most atomic way to train and inference a GPT in pure, dependency-free JavaScript.
 * This file is the complete algorithm.
 * Everything else is just efficiency.
 *
 * @karpathy (original Python), converted to JavaScript (Node.js)
 */

const fs = require("fs");

// --- RNG (xoshiro256** seeded with 42, using BigInt for 64-bit) ---
class Rng {
  constructor(seed) {
    // SplitMix64 to seed xoshiro
    let sm = BigInt(seed);
    const mask = (1n << 64n) - 1n;
    const smNext = () => {
      sm = (sm + 0x9e3779b97f4a7c15n) & mask;
      let z = sm;
      z = ((z ^ (z >> 30n)) * 0xbf58476d1ce4e5b9n) & mask;
      z = ((z ^ (z >> 27n)) * 0x94d049bb133111ebn) & mask;
      return (z ^ (z >> 31n)) & mask;
    };
    this.s = [smNext(), smNext(), smNext(), smNext()];
    this._mask = mask;
  }

  _rotl(x, k) {
    const m = this._mask;
    return ((x << k) | (x >> (64n - k))) & m;
  }

  nextU64() {
    const m = this._mask;
    const s = this.s;
    const result = (this._rotl((s[1] * 5n) & m, 7n) * 9n) & m;
    const t = (s[1] << 17n) & m;
    s[2] ^= s[0];
    s[3] ^= s[1];
    s[1] ^= s[2];
    s[0] ^= s[3];
    s[2] ^= t;
    s[3] = this._rotl(s[3], 45n);
    return result;
  }

  nextF64() {
    return Number(this.nextU64() >> 11n) / 2 ** 53;
  }

  gauss(mean, std) {
    const u1 = Math.max(this.nextF64(), 1e-30);
    const u2 = this.nextF64();
    return mean + std * Math.sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.PI * u2);
  }

  shuffle(arr) {
    for (let i = arr.length - 1; i > 0; i--) {
      const j = Number(this.nextU64() % BigInt(i + 1));
      [arr[i], arr[j]] = [arr[j], arr[i]];
    }
  }

  weightedChoice(weights) {
    const total = weights.reduce((a, b) => a + b, 0);
    let r = this.nextF64() * total;
    for (let i = 0; i < weights.length; i++) {
      r -= weights[i];
      if (r <= 0) return i;
    }
    return weights.length - 1;
  }
}

// --- Let there be Autograd ---
class Value {
  constructor(data, children = [], localGrads = []) {
    this.data = data;           // scalar value of this node calculated during forward pass
    this.grad = 0.0;            // derivative of the loss w.r.t. this node
    this.children = children;   // children of this node in the computation graph
    this.localGrads = localGrads; // local derivative of this node w.r.t. its children
  }

  add(other) {
    other = other instanceof Value ? other : new Value(other);
    return new Value(this.data + other.data, [this, other], [1.0, 1.0]);
  }

  mul(other) {
    other = other instanceof Value ? other : new Value(other);
    return new Value(this.data * other.data, [this, other], [other.data, this.data]);
  }

  pow(exp) {
    return new Value(Math.pow(this.data, exp), [this], [exp * Math.pow(this.data, exp - 1)]);
  }

  log() {
    return new Value(Math.log(this.data), [this], [1.0 / this.data]);
  }

  exp() {
    const e = Math.exp(this.data);
    return new Value(e, [this], [e]);
  }

  relu() {
    return new Value(Math.max(0, this.data), [this], [this.data > 0 ? 1.0 : 0.0]);
  }

  neg() { return this.mul(-1); }
  sub(other) { other = other instanceof Value ? other : new Value(other); return this.add(other.neg()); }
  div(other) { other = other instanceof Value ? other : new Value(other); return this.mul(other.pow(-1)); }

  backward() {
    const topo = [];
    const visited = new Set();
    const buildTopo = (v) => {
      if (visited.has(v)) return;
      visited.add(v);
      for (const child of v.children) buildTopo(child);
      topo.push(v);
    };
    buildTopo(this);
    this.grad = 1.0;
    for (let i = topo.length - 1; i >= 0; i--) {
      const v = topo[i];
      for (let j = 0; j < v.children.length; j++) {
        v.children[j].grad += v.localGrads[j] * v.grad;
      }
    }
  }
}

// --- Utility ---
function sumVals(arr) {
  let s = new Value(0);
  for (const x of arr) s = s.add(x);
  return s;
}

// --- Model functions ---
function linear(x, w) {
  return w.map((wo) => {
    let s = new Value(0);
    for (let i = 0; i < wo.length; i++) s = s.add(wo[i].mul(x[i]));
    return s;
  });
}

function softmax(logits) {
  const maxVal = Math.max(...logits.map((v) => v.data));
  const exps = logits.map((v) => v.sub(maxVal).exp());
  const total = sumVals(exps);
  return exps.map((e) => e.div(total));
}

function rmsnorm(x) {
  let ms = new Value(0);
  for (const xi of x) ms = ms.add(xi.mul(xi));
  ms = ms.div(x.length);
  const scale = ms.add(1e-5).pow(-0.5);
  return x.map((xi) => xi.mul(scale));
}

// --- Main ---
(function main() {
  const rng = new Rng(42);

  // Let there be an input dataset
  if (!fs.existsSync("input.txt")) {
    console.error("Error: input.txt not found. Please download it first.");
    process.exit(1);
  }
  const content = fs.readFileSync("input.txt", "utf-8");
  const docs = content.split("\n").map((l) => l.trim()).filter((l) => l.length > 0);
  rng.shuffle(docs);
  console.log(`num docs: ${docs.length}`);

  // Let there be a Tokenizer
  const charSet = new Set();
  for (const d of docs) for (const c of d) charSet.add(c);
  const uchars = [...charSet].sort();
  const BOS = uchars.length;
  const vocabSize = uchars.length + 1;
  console.log(`vocab size: ${vocabSize}`);

  const charIndex = (c) => uchars.indexOf(c);

  // Initialize the parameters
  const nEmbd = 16;
  const nHead = 4;
  const nLayer = 1;
  const blockSize = 16;
  const headDim = nEmbd / nHead;

  const makeMatrix = (nout, nin) =>
    Array.from({ length: nout }, () =>
      Array.from({ length: nin }, () => new Value(rng.gauss(0, 0.08)))
    );

  const wte = makeMatrix(vocabSize, nEmbd);
  const wpe = makeMatrix(blockSize, nEmbd);
  const lmHead = makeMatrix(vocabSize, nEmbd);

  const layers = [];
  for (let i = 0; i < nLayer; i++) {
    layers.push({
      attnWq: makeMatrix(nEmbd, nEmbd),
      attnWk: makeMatrix(nEmbd, nEmbd),
      attnWv: makeMatrix(nEmbd, nEmbd),
      attnWo: makeMatrix(nEmbd, nEmbd),
      mlpFc1: makeMatrix(4 * nEmbd, nEmbd),
      mlpFc2: makeMatrix(nEmbd, 4 * nEmbd),
    });
  }

  // Flatten params
  const params = [];
  for (const row of wte) for (const p of row) params.push(p);
  for (const row of wpe) for (const p of row) params.push(p);
  for (const row of lmHead) for (const p of row) params.push(p);
  for (const lw of layers) {
    for (const mat of [lw.attnWq, lw.attnWk, lw.attnWv, lw.attnWo, lw.mlpFc1, lw.mlpFc2]) {
      for (const row of mat) for (const p of row) params.push(p);
    }
  }
  console.log(`num params: ${params.length}`);

  // GPT function
  function gpt(tokenId, posId, keys, values) {
    const tokEmb = wte[tokenId];
    const posEmb = wpe[posId];
    let x = tokEmb.map((t, i) => t.add(posEmb[i]));
    x = rmsnorm(x);

    for (let li = 0; li < nLayer; li++) {
      // 1) Multi-head attention block
      const xResidual = [...x];
      x = rmsnorm(x);
      const q = linear(x, layers[li].attnWq);
      const k = linear(x, layers[li].attnWk);
      const v = linear(x, layers[li].attnWv);
      keys[li].push(k);
      values[li].push(v);
      const xAttn = [];
      for (let h = 0; h < nHead; h++) {
        const hs = h * headDim;
        const qH = q.slice(hs, hs + headDim);
        const kH = keys[li].map((ki) => ki.slice(hs, hs + headDim));
        const vH = values[li].map((vi) => vi.slice(hs, hs + headDim));
        const scale = Math.sqrt(headDim);
        const attnLogits = kH.map((kt) => {
          let dot = new Value(0);
          for (let j = 0; j < headDim; j++) dot = dot.add(qH[j].mul(kt[j]));
          return dot.div(scale);
        });
        const attnWeights = softmax(attnLogits);
        for (let j = 0; j < headDim; j++) {
          let s = new Value(0);
          for (let t = 0; t < vH.length; t++) s = s.add(attnWeights[t].mul(vH[t][j]));
          xAttn.push(s);
        }
      }
      x = linear(xAttn, layers[li].attnWo);
      x = x.map((xi, i) => xi.add(xResidual[i]));
      // 2) MLP block
      const xResidual2 = [...x];
      x = rmsnorm(x);
      x = linear(x, layers[li].mlpFc1);
      x = x.map((xi) => xi.relu());
      x = linear(x, layers[li].mlpFc2);
      x = x.map((xi, i) => xi.add(xResidual2[i]));
    }

    return linear(x, lmHead);
  }

  // Let there be Adam
  const learningRate = 0.01;
  const beta1 = 0.85;
  const beta2 = 0.99;
  const epsAdam = 1e-8;
  const mBuf = new Float64Array(params.length);
  const vBuf = new Float64Array(params.length);

  // Repeat in sequence
  const numSteps = 1000;
  for (let step = 0; step < numSteps; step++) {
    // Take single document, tokenize it
    const doc = docs[step % docs.length];
    const tokens = [BOS];
    for (const c of doc) tokens.push(charIndex(c));
    tokens.push(BOS);
    const n = Math.min(blockSize, tokens.length - 1);

    // Forward
    const keys = Array.from({ length: nLayer }, () => []);
    const vals = Array.from({ length: nLayer }, () => []);
    const losses = [];
    for (let posId = 0; posId < n; posId++) {
      const tokenId = tokens[posId];
      const targetId = tokens[posId + 1];
      const logits = gpt(tokenId, posId, keys, vals);
      const probs = softmax(logits);
      const lossT = probs[targetId].log().neg();
      losses.push(lossT);
    }
    const loss = sumVals(losses).div(n);

    // Backward
    loss.backward();

    // Adam optimizer update
    const lrT = learningRate * (1 - step / numSteps);
    for (let i = 0; i < params.length; i++) {
      const p = params[i];
      mBuf[i] = beta1 * mBuf[i] + (1 - beta1) * p.grad;
      vBuf[i] = beta2 * vBuf[i] + (1 - beta2) * p.grad * p.grad;
      const mHat = mBuf[i] / (1 - Math.pow(beta1, step + 1));
      const vHat = vBuf[i] / (1 - Math.pow(beta2, step + 1));
      p.data -= lrT * mHat / (Math.sqrt(vHat) + epsAdam);
      p.grad = 0;
    }

    const s1 = String(step + 1).padStart(4);
    const s2 = String(numSteps).padStart(4);
    console.log(`step ${s1} / ${s2} | loss ${loss.data.toFixed(4)}`);
  }

  // Inference
  const temperature = 0.5;
  console.log("\n--- inference (new, hallucinated names) ---");
  for (let sampleIdx = 0; sampleIdx < 20; sampleIdx++) {
    const keys = Array.from({ length: nLayer }, () => []);
    const vals = Array.from({ length: nLayer }, () => []);
    let tokenId = BOS;
    const sample = [];
    for (let posId = 0; posId < blockSize; posId++) {
      const logits = gpt(tokenId, posId, keys, vals);
      const tempLogits = logits.map((l) => l.div(temperature));
      const probs = softmax(tempLogits);
      const weights = probs.map((p) => p.data);
      tokenId = rng.weightedChoice(weights);
      if (tokenId === BOS) break;
      sample.push(uchars[tokenId]);
    }
    const idx = String(sampleIdx + 1).padStart(2);
    console.log(`sample ${idx}: ${sample.join("")}`);
  }
})();
